### ISSUE TITLE
Visual Studio should warn if an analyzer targets .NET Framework

### ISSUE THREAD

**Issue by jaredpar** [dotnet/roslyn#44251]
From their inception the intent of analyzers was for them to build against portable target frameworks. Originally this was various portable profiles and more recently it's `netstandard2.0`. The reasoning is that analyzers must load in the compiler and the compiler runs on a variety of runtimes: mono, .NET Framework and .NET Core. By targeting `netstandard2.0` analyzers, and now source generators, ensure they can work in all situations.

This guideline though was never enforced. That means analyzers that targeted .NET Framework and were used entirely within the .NET Framework tools, essentially through Visual Studio, would have likely worked without issue. 

Taking a step back though this is going to present some challenges for the overall ecosystem:

1. Such analyzers will not work when the build is driven via the .NET SDK because the compiler runs on .NET Core.
1. Such analyzers will fail in all scenarios when the compiler shifts to being a .NET Core only tool. 

To begin pushing analyzer authors to target a portable profile we should be issuing a warning, diagnostic, yellow bar, etc ... in Visual Studio when an analyzer is loaded that targets .NET Framework. This will start alerting analyzer authors to the problem and we can then work with them to move to target `netstandard2.0`

**Comment by sharwell** [comment]
> 1. Such analyzers will not work when the build is driven via the .NET SDK because the compiler runs on .NET Core.
> 1. Such analyzers will fail in all scenarios when the compiler shifts to being a .NET Core only tool.

This is perhaps a bit strong. The analyzers _might_ not work, but if the analyzer doesn't do anything out of the ordinary it may continue to work the same in .NET Core as it did on .NET Framework.

**Comment by tmat** [comment]
I believe reporting a IDE-only warning is better then just crossing fingers and hoping the analyzer will work.

**Comment by jaredpar** [comment]
I agree that there are definitely cases where a .NET Framework targeted analyzer can load and execute on .NET Core (and vice versa). The .NET Core runtime actually supports this to a degree with a number of libraries. 

I would say though that for the specific case of analyzers this is almost certainly not an explicit design decision.. Rather the author just chose a .NET Framework TFM that they were familiar with.

**Comment by sharwell** [comment]
Design review conclusion: if an analyzer can help us figure out why an analyzer has started failing when it ran on .NET Core instead of .NET Framework, then the analyzer would be valuable to the team. At minimum, we want to make sure the analyzer is available (and running) if/when the following occurs:

1. The IDE is using .NET Core for execution of analyzers in a project that only targets .NET Framework (e.g. if we move Roslyn OOP to run on .NET Core instead of .NET Framework).
2. The command line build uses .NET Framework for MSBuild execution, but the compiler is running through VBCSCompiler.exe which is running under .NET Core.

We're interested in ideas the compiler team may have regarding a design that works in the core analyzer driver, but if it's not possible to have a feature there automatically (for compat reasons) then we can implement it as an IDE-specific feature.

**Comment by jaredpar** [comment]
>  if an analyzer can help us figure out why an analyzer has started failing when it ran on .NET Core instead of .NET Framework, then the analyzer would be valuable to the team

I don't understand why the conversation ended up going this route vs. the more straight forward warn if an analyzer targets .NET Framework. That is an unsupported scenario. Having an analyzer that helps diagnose why an unsupported scenario doesn't work seems like it's not a good use of resources especially when detecting the supported scenarios is straight forward.

**Comment by sharwell** [comment]
> warn if an analyzer targets .NET Framework

Yes, this is the analyzer we were discussing.
